doc/plugin_agent_nodeattestor_tpm_devid.md

```diff
@@ -0,0 +1,58 @@
+# Agent plugin: NodeAttestor "tpm_devid"
+
+*Must be used in conjunction with the server-side tpm_devid plugin*
+
+The `tpm_devid` plugin provides attestation data for a node that owns a TPM
+and that has been provisioned with a DevID certificate through an out-of-band 
+mechanism.
+
+The plugin responds to two challenges requested by the server:
+
+1. A proof-of-possession challenge: The agent receives and signs a random nonce 
+to prove it is in possession of the private key that corresponds to the DevID 
+certificate presented to the server.
+
+2. A proof-of-residency challenge: The agent receives and solves a 
+specially-crafted, encrypted challenge to prove to the server that the DevID 
+keypair was generated and resides in a TPM of a trusted vendor.
+
+
+The SPIFFE ID produced by the server-side `tpm_devid` plugin is based on the 
+DevID certificate fingerprint, where the fingerprint is defined as the SHA1 hash
+of the ASN.1 DER encoding of the identity certificate.
+
+The SPIFFE ID has the form:
+
+```
+spiffe://<trust domain>/spire/agent/tpm_devid/<fingerprint>
+```
+
+| Configuration 	| Description 													| Default	|
+| ---------------- 	| -------------------------------------------------------------	| ---------	|
+|`tpm_device_path`	| The path to a TPM 2.0 device.									|     ""	|
+|`devid_cert_path`	| The path to the DevID certificate on disk in PEM format.		|			|
+|`devid_priv_path`	| The path to the private key blob generated by the TPM.		|			|
+|`devid_pub_path` 	| The path to the public key blob generated by the TPM.			|			|
+
+
+A sample configuration:
+
+```
+	NodeAttestor "tpm_devid" {
+		plugin_data {
+            tpm_device_path = "/dev/tpmrm0"
+			devid_cert_path = "/opt/spire/conf/agent/devid.crt.pem"
+            devid_priv_path = "/opt/spire/conf/agent/devid.priv.blob"
+			devid_pub_path = "/opt/spire/conf/agent/devid.pub.blob"
+		}
+	}
+```
+
+If `tpm_device_path` is unset, the plugin will try to auto-detect the TPM in
+the node.
+
+### Compatibility considerations
+
++ This plugin is designed to work with TPM 2.0, TPM 1.2 is not supported.
++ Only local device identities (LDevIDs) are supported. Attestation using 
+IDevIDs is not supported.
```

Comment:
> Is it not useful to attest using IDevID?

Yes, it is useful to attest nodes using IDevIDs. The plugin is currently restricted to use LDevID identities with ordinary keys in order to reduce the scope of the implementation.

There are two types of TPM-based keys: primary keys and ordinary keys.
 
A primary key is created by the TPM based on the current primary seed and, unless persisted within the TPM, it must be recreated before it can be used.

An ordinary key is produced with a seed taken from the TPM random number generator. An ordinary key may be persisted within the TPM or it may be persisted externally to the TPM in the form of an encrypted key blob.
 
According to the [TCG draft spec (section 3.7)](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_DevID_v1r2_02dec2020.pdf), IDevID keys must be primary keys. So, If we decide to support IDevIDs (now or in the future) we would have to update the code to load/regenerate/get the keys differently into the TPM depending on if it is an ordinary key or a primary key.

> What kind of information is required to be in an IDevID, and what kind of information is found there in practice?

In general, LDevIDs and IDevIDs certificates contains similar information with different "MUST"/"SHOULD" requirements. Some of the data that should be present is: the device serial number, hardware module name, hardware serial number, etc. The [TCG spec (section 8.1)](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_DevID_v1r2_02dec2020.pdf) enumerates the certificate fields and expected information for TPM-based (I/L)DevIDs.